File: Nov2020/25Nov/AddTwoNumbersII.java
/*
Problem Name: Add Two Numbers II
Problem Link: https://leetcode.com/problems/add-two-numbers-ii/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        List<Integer> list1 = new ArrayList<Integer>();
        List<Integer> list2 = new ArrayList<Integer>();
        while(l1 != null){
            list1.add(0, l1.val);
            l1 = l1.next;
        }
        while(l2 != null){
            list2.add(0, l2.val);
            l2 = l2.next;
        }
        List<Integer> ans = new ArrayList<Integer>();
        int size1 = list1.size(), size2 = list2.size(), n = Math.max(size1,size2), carry = 0;
        for(int i=0;i<n;i++){
            int sum = carry;
            if(i < size1)
                sum += list1.get(i);
            if(i < size2)
                sum += list2.get(i);
            ans.add(sum%10);
            carry = sum/10;
        }
        if(carry > 0)
            ans.add(carry);
        ListNode ansLL = new ListNode(-1), trav = ansLL;
        for(int i=ans.size()-1;i>=0;i--){
            trav.next = new ListNode(ans.get(i));
            trav = trav.next;
        }
        return ansLL.next;
    }
}

File: Nov2020/25Nov/AddTwoNumbersII.cpp
/*
Problem name: 445. Add Two Numbers II
Problem link: https://leetcode.com/problems/add-two-numbers-ii/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        stack<int> s1,s2;
        ListNode* temp;
        int x,y,num=0,carry=0,rem=0;
        
        temp=l1;
        while(temp!=NULL){
            s1.push(temp->val);
            temp=temp->next;
        }
        temp=l2;
        while(temp!=NULL){
            s2.push(temp->val);
            temp=temp->next;
        }
        
        ListNode* ans=NULL;
        while(!s1.empty() || !s2.empty()){
            x=0;
            y=0;
            if(!s1.empty()){
                x=s1.top();
                s1.pop();
            }
            if(!s2.empty()){
                y=s2.top();
                s2.pop();
            }
            num=x+y+carry;
            carry=num/10;
            rem=num%10;
            
            ListNode* q=new ListNode(rem);
            q->next=ans;
            ans=q;
        }
        if(carry!=0){
            ListNode* q=new ListNode(carry);
            q->next=ans;
            ans=q;
        }
        return ans;
    }
};
